src/clients/patreon/patreonAgent.ts

```diff
@@ -0,0 +1,235 @@
+import patreonId from '../../models/patreonId';
+import { PatreonController, PatreonUserTokens } from './patreonController';
+
+interface PatreonDetails {
+  patreonUserId: string;
+  patreonUsersName: string;
+  isActivePatreon: boolean;
+  amountCents: number;
+}
+
+export interface IPatreonController {
+  getTokens(code: string): Promise<PatreonUserTokens>;
+  // TODO-kev: Change the any type
+  getPatronDetails(accessToken: string): any;
+  getLoginUrl(): string;
+}
+
+export class PatreonAgent {
+  private patreonController: IPatreonController;
+
+  constructor(controller: IPatreonController) {
+    this.patreonController = controller;
+  }
+
+  public getPatreonAuthorizationUrl() {
+    return this.patreonController.getLoginUrl();
+  }
+
+  public async getExistingPatronDetails(
+    usernameLower: string,
+  ): Promise<PatreonDetails> {
+    // This function is to check for features in general on load
+
+    const existingPatreon = await this.getExistingPatreon(usernameLower);
+
+    if (!existingPatreon) {
+      return null;
+    }
+
+    const isActivePatreon = !this.hasExpired(
+      existingPatreon.currentPledgeExpiryDate,
+    );
+
+    return {
+      patreonUserId: existingPatreon.patreonUserId,
+      patreonUsersName: existingPatreon.patreonUsersName,
+      isActivePatreon,
+      amountCents: existingPatreon.amountCents,
+    };
+  }
+
+  // This path is hit whenever user clicks link to Patreon button
+  public async linkUserToPatreon(
+    usernameLower: string,
+    code: string,
+  ): Promise<PatreonDetails> {
+    // Grab user tokens
+    const tokens = await this.patreonController.getTokens(code);
+
+    // Grab member details from Patreon with token
+    const patronDetails = await this.patreonController.getPatronDetails(
+      tokens.userAccessToken,
+    );
+
+    // Grab Patreon document from MongoDB
+    const existingPatreon = await this.getExistingPatreon(usernameLower);
+
+    // Do not let more than one patreon be used for same user
+    if (
+      existingPatreon &&
+      existingPatreon.patreonUserId !== patronDetails.patreonUserId
+    ) {
+      throw new Error(
```

Comment:
Can you add a comment saying we don't ever expect this to be hit as the GUI should enforce a user to not be able to link to another patreon while they're still linked to another one.